### ISSUE TITLE
What's the replacement for http_uri:encode/1 ?

### ISSUE THREAD

**Issue by elbrujohalcon** [erlang/otp#4884]
Since OTP23, `http_uri:encode/1` is _deprecated_.
If you write code using it, the compiler will tell you that you should [_use `uri_string` functions instead_](https://github.com/erlang/otp/blob/d256ae477014158a49bb860b283df9c040011197/lib/inets/src/http_lib/http_uri.erl#L67).
But I didn't find any function in `uri_string` with which I could directly replace `http_uri:encode/1`.
For instance, I was looking for a function to produce this output…

```erlang
2> http_uri:encode("use:percents").
"use%3Apercents"
```

I looked up within OTP code itself for a place where `http_uri:encode/1` was used before (in OTP22) to see what you replaced it with. But, to my surprise, I only found [`httpd_util:encode_hex/1`](https://github.com/erlang/otp/blob/d256ae477014158a49bb860b283df9c040011197/lib/inets/src/http_server/httpd_util.erl#L401-L402), where `http_uri:encode/1` is **still used**. Instead of replacing it with functions from `uri_string`, [these lines](https://github.com/erlang/otp/commit/94c7ae3b32dd9db760b974acc26dc47b874ee121) were added to `httpd_util`.

So, my question is: If I don't want to disable warnings in my code… Exactly what `uri_string` functions I should use to replace `http_uri:encode/1`?

**Comment by lhoguin** [comment]
What percent encoding do you need? There's several depending on what part of the URI you are encoding. Can you tell us what you are encoding exactly so we can tell you what the `uri_string` equivalent is?

**Comment by elbrujohalcon** [comment]
Yeah. I was actually upgrading an open-source repository so I can show you exactly what I need to change: https://github.com/AdRoll/erliam/pull/18#discussion_r637955229

The things that goes through that code are AWS IAM role names (stuff like `arn:aws:iam::123456789012:role/service-role/QuickSightAction`) and they need to be _percent-encoded_ (i.e. `arn%3Aaws%3Aiam%3A%3A123456789012%3Arole%2Fservice-role%2FQuickSightAction`).

**Comment by lhoguin** [comment]
After a quick look, I don't see any mention of encoding in Authorization header (https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html) or in request bodies. The only mention I see is here (https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html) where you can pass it in the query string. I know it's not IAM docs per say so perhaps it's not the same there but it seems to match what's in the code in awsv4.erl. Anyway in that case you would need query string url encoding. But that doesn't seem to be what you want since you want slashes encoded. So I don't know. :-)

**Comment by elbrujohalcon** [comment]
Yeah. I'm not the original author of `erliam` (that's @zerth) but… I'm 100% willing to trust that if he put that encoding there, we _definitely_ needed it for some reason. 🤷🏻 
In any case, even if in this particular scenario I don't need to encode anything, at some point I'll likely need the reverse operation of `uri_string:percent_decode/1`, anyway.
As a matter of fact, the OTP will _soon_ need it, too… Since `http_uri:encode/1` is scheduled for removal on OTP25 and then somebody will need to fix `httpd_util:encode_hex/1` implementation, right?

**Comment by lhoguin** [comment]
So far similar tickets have shown the need for a percent_decode (it exists because it is necessary to decode path segments for example since the other functions can't convert the %2f back to slashes when it returns the full path) but not for a percent encode.

**Comment by IngelaAndin** [comment]
> Yeah. I'm not the original author of `erliam` (that's @zerth) but… I'm 100% willing to trust that if he put that encoding there, we _definitely_ needed it for some reason. 🤷🏻
> In any case, even if in this particular scenario I don't need to encode anything, at some point I'll likely need the reverse operation of `uri_string:percent_decode/1`, anyway.
> As a matter of fact, the OTP will _soon_ need it, too… Since `http_uri:encode/1` is scheduled for removal on OTP25 and then somebody will need to fix `httpd_util:encode_hex/1` implementation, right?

Not really as httpd_util:encode_hex/1 is dead code and not documented.  There are a lot legacy stuff in the inets application that is only almost what you want!  We are trying to phase out as much as possible and improve what must remain for now.  We will see what happens in the long run.  

You should read chapter 4 of the stdlib Users Guide.

**Comment by elbrujohalcon** [comment]
So, @IngelaAndin … you're going to remove `httpd_util` before OTP25, then?

**Comment by zerth** [comment]
> Yeah. I'm not the original author of erliam (that's @zerth) but… I'm 100% willing to trust that if he put that encoding there, we definitely needed it for some reason. 🤷🏻

`erliam` is the result of factoring out common behavior from `dinerl` and `kinetic`, but it's entirely possible I introduced something not actually needed during that process.
For this case, I would expect you could just use the new normalization function on a `uri_map()` containing the entire path of interest (and whatever other components) and directly use the result.

**Comment by u3s** [comment]
@elbrujohalcon can you let us know what is the status of that issue from your perspective? I saw you considered replacing `http_uri:encode/1` calls with `httpd_util:encode_hex/1`, but when I checked recent branch of `erliam` I see no more references to any of these functions.

also noted `awsv4` module contains functionality that probably does what you need.

if you are not relying on that functionality anymore, can this ticket be closed?

**Comment by u3s** [comment]
FYI, discussion about replacement is taking place in #5368 .
